src/hosts.rs

```
@@ -0,0 +1,323 @@
+    /// This is done by querying the SHOW READYSET STATUS command.
+    ///
+    /// # Returns
+    ///
+    /// true if the host is ready, false otherwise.
+    pub fn check_readyset_is_ready(&mut self) -> Result<bool, mysql::Error> {
+        match &mut self.conn {
+            Some(conn) => {
+                let rows: Vec<(String, String)> =
+                    conn.query("SHOW READYSET STATUS").unwrap_or(vec![]);
+                for (field, value) in rows {
+                    if field == "Snapshot Status" {
+                        return Ok(value == "Completed");
+                    }
+                }
+                Ok(false)
+            }
+            None => Err(mysql::Error::IoError(std::io::Error::new(
+                std::io::ErrorKind::Other,
+                "Connection to Readyset host is not established",
+            ))),
+        }
+    }
+
+    /// Checks if the host supports the given query.
+    /// This is done by querying the EXPLAIN CREATE CACHE FROM command.
+    ///
+    /// # Arguments
+    ///
+    /// * `digest_text` - The digest text of the query.
+    /// * `schema` - The schema of the query.
+    ///
+    /// # Returns
+    ///
+    /// true if the host supports the query, false otherwise.
+    pub fn check_query_support(
+        &mut self,
+        digest_text: &String,
+        schema: &String,
+    ) -> Result<bool, mysql::Error> {
+        match &mut self.conn {
+            Some(conn) => {
+                conn.query_drop(format!("USE {}", schema))
+                    .expect("Failed to use schema");
+                let row: Option<(String, String, String)> =
+                    conn.query_first(format!("EXPLAIN CREATE CACHE FROM {}", digest_text))?;
+                match row {
+                    Some((_, _, value)) => Ok(value == "yes" || value == "cached"),
+                    None => Ok(false),
+                }
+            }
+            None => Ok(false),
```

Hmm, why not return an error here?

---

I don't know one example of when EXPLAIN will return no rows. I am running the query with `?` which will make the caller return the same error. I think it still make sense to differ an Error from an empty result.